builtins/amp-img.js

```diff
@@ -21,72 +21,103 @@ import {registerElement} from '../src/custom-element';
 
-    /** @override */
-    isLayoutSupported(layout) {
-      return isLayoutSizeDefined(layout);
+    if (this.element.id) {
+      this.img_.setAttribute('amp-img-id', this.element.id);
     }
+    this.propagateAttributes(['alt'], this.img_);
+    this.applyFillContent(this.img_, true);
 
-    /** @override */
-    buildCallback() {
-      /** @private @const {!Element} */
-      this.img_ = new Image();
+    this.img_.width = getLengthNumeral(this.element.getAttribute('width'));
+    this.img_.height = getLengthNumeral(this.element.getAttribute('height'));
 
-      if (this.element.id) {
-        this.img_.setAttribute('amp-img-id', this.element.id);
-      }
-      this.propagateAttributes(['alt'], this.img_);
-      this.applyFillContent(this.img_, true);
+    this.element.appendChild(this.img_);
 
-      this.img_.width = getLengthNumeral(this.element.getAttribute('width'));
-      this.img_.height = getLengthNumeral(this.element.getAttribute('height'));
+    /** @private @const {!Srcset} */
+    this.srcset_ = srcsetFromElement(this.element);
+  }
 
-      this.element.appendChild(this.img_);
+  /** @override */
+  prerenderAllowed() {
+    return true;
+  }
 
-      /** @private @const {!Srcset} */
-      this.srcset_ = srcsetFromElement(this.element);
-    }
+  /** @override */
+  isRelayoutNeeded() {
+    return true;
+  }
 
-    /** @override */
-    prerenderAllowed() {
-      return true;
-    }
+  /** @override */
+  layoutCallback() {
+    let promise = this.updateImageSrc_();
 
-    /** @override */
-    isRelayoutNeeded() {
-      return true;
+    // we only allow to fallback on error on the initial layoutCallback
+    // or else this would be pretty expensive.
+    if (!this.allowImgLoadFallback_) {
+      promise = promise.catch(this.onImgLoadingError_.bind(this));
+      this.allowImgLoadFallback_ = true;
     }
-
-    /** @override */
-    layoutCallback() {
-      return this.updateImageSrc_();
+    return promise;
+  }
+
+  /**
+   * @return {!Promise}
+   * @private
+   */
+  updateImageSrc_() {
+    if (this.getLayoutWidth() <= 0) {
+      return Promise.resolve();
+    }
+    const src = this.srcset_.select(this.getLayoutWidth(), this.getDpr()).url;
+    if (src == this.img_.getAttribute('src')) {
+      return Promise.resolve();
     }
 
-    /**
-     * @return {!Promise}
-     * @private
-     */
-    updateImageSrc_() {
-      if (this.getLayoutWidth() <= 0) {
-        return Promise.resolve();
-      }
-      const src = this.srcset_.select(this.getLayoutWidth(), this.getDpr()).url;
-      if (src == this.img_.getAttribute('src')) {
-        return Promise.resolve();
-      }
-      this.img_.setAttribute('src', src);
-
-      return loadPromise(this.img_);
+    // clean up the fallback if the src has changed.
+    if (this.allowImgLoadFallback_ &&
+        this.img_.classList.contains('-amp-ghost')) {
+      this.img_.classList.remove('-amp-ghost');
+      this.toggleFallback(false);
```

Comment:
done. added it on the promise returned from the `updateImgSrc_` funtion, can't add it on the layoutCallback as we can't distinguish if it a promise created from `loadPromise` (we actually do a fetch) or a promise created from `Promise.resolve()` which makes the `updateImgSrc_` function a noop.